### PR TITLE
Adding SIP routing functionality to PhoneNumbers SDK

### PR DESCRIPTION
**Pull request by jiriburant**
# Description

Added the SipRoutingClient enabling management of SIP routing configuration for `direct routing` numbers.
Each Azure Communication resource has exactly one SIP configuration, containing a dictionary of trunks and a list of routes.
The trunks allow for partial updates, the configuration is merged with the incoming trunk dictionary. The routes don’t allow for partial updates, the whole list is replaced by the incoming list (if it is present on PATCH request).

The SIP routing client provides straigthforward methods to retrieve and manage SIP trunks and voice routes.

Note:
The API we are calling from the SDK is rather different from the SDK public interface, so we had to introduce two types for SIP trunks, called "SipTrunk" and "SipTrunkInternal", where "SipTrunk" is exposed to user and "SipTrunkInternal" is used for communication with backend API. This can be a bit confusing in the code, when the info is mapped to and from those types.
 
Resources:

API specification with API examples:   https://github.com/Azure/azure-rest-api-specs/tree/main/specification/communication/data-plane/SipRouting/preview/2021-05-01-preview
Conceptual doc, explaining the direct routing provisioning scenario:  https://docs.microsoft.com/en-us/azure/communication-services/concepts/telephony/direct-routing-provisioning
Apiview link: https://apiview.dev/Assemblies/Review/b32afd45c457422992f506154f376d68

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/communication/azure-communication-phonenumbers/README.md [hunk 1]
```diff
@@ -143,6 +182,46 @@ poller = self.phone_number_client.begin_update_phone_number_capabilities(
...
+routes = sip_routing_client.get_routes()
+```
+
+#### Replace SIP trunks and routes
+
+Replace the list of currently configured trunks or routes with new values.
+
+```python
+sip_routing_client.set_trunks("<list of new trunks>")
+sip_routing_client.set_routes("<list of new routes>")
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/README.md:202]: Rather than using a string as the placeholder input - maybe it should be a list to avoid any confusion?
Better yet - an actual example of what this input might look like.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/__init__.py [hunk 2]
```diff
@@ -4,6 +4,8 @@
 # license information.
 # --------------------------------------------------------------------------
 
+__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/__init__.py:7]: Why was this added?

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/__init__.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/__init__.py:1]: This line shouldn't be here.
Additionally this file seems to be missing the license header.

File: sdk/communication/azure-communication-phonenumbers/README.md [hunk 4]
```diff
@@ -143,6 +182,46 @@ poller = self.phone_number_client.begin_update_phone_number_capabilities(
...
 
+### SipRoutingClient
+
+#### Retrieve SIP trunks and routes
+
+Get the list of currently configured trunks or routes.
+
+```python
+trunks = sip_routing_client.get_trunks()
+routes = sip_routing_client.get_routes()
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/README.md:193]: Ideally for the code snippets we would actually demonstrate some usage:
```python
trunks = client.get_trunks()
for trunk in trunks:
    print(trunk.sip_signaling_port)
```

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    :keyword api_version: Api Version. Default value is "2021-05-01-preview". Note that overriding
+     this default value may result in unsupported behavior.
+    :paramtype api_version: str
+    """
+
+    def __init__(
+        self,
+        endpoint,  # type: str
+        credential,  # type: TokenCredential
+        api_version="2021-05-01-preview", # type: str
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py:51]: This param shouldn't be positional - intead we can just `pop` it from kwargs.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        """
+        if trunk_fqdn is None:
+            raise ValueError("Parameter 'trunk_fqdn' must not be None.")
+
+        config = self._rest_service.get_sip_configuration(
+            **kwargs)
+        trunk = config.trunks[trunk_fqdn]
+
+        if not trunk:
+            raise LookupError("No entry found for FQDN:" + trunk_fqdn)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py:116]: I don't think we should be raising a `LookupError` here.
I'm not really a fan of raising in this scenario. Is that what's being done across languages?

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        self._rest_service.patch_sip_configuration(
+            body=SipConfiguration(trunks={trunk_fqdn:None}),
+            **kwargs)
+
+    @distributed_trace
+    def get_trunks(
+        self,
+        **kwargs  # type: Any
+    ):  # type: (...) -> List[SipTrunk]
+        """Getter for currently configured SIP trunks.
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py:165]: 'Getter' sounds very Java ;)
Maybe just 'Retrieves an iterable of currently configured SIP trunks."?

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+        self._rest_service.patch_sip_configuration(
+            body=SipConfiguration(trunks={trunk_fqdn:None}),
+            **kwargs)
+
+    @distributed_trace
+    def get_trunks(
+        self,
+        **kwargs  # type: Any
+    ):  # type: (...) -> List[SipTrunk]
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py:164]: We should document this as `Iterable[SipTrunk]`
It gives us more flexibility in future if we change the return type.

Comment (annatisch) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py:164]: Same applies to `get_routes()`

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/aio/_sip_routing_client_async.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+    async def close(self) -> None:
+        await self._rest_service.close()
+
+    async def __aenter__(self) -> "SipRoutingClient":
+        await self._rest_service.__aenter__()
+        return self
+
+    async def __aexit__(self, *args) -> None:
+        await self._rest_service.__aexit__(*args)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/aio/_sip_routing_client_async.py:267]: Could we please add the `close`/`__enter__`/`__exit__` overrides to the sync client as well?

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/delete_sip_trunk_sample_async.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+"""
+
+import os
+import asyncio
+from azure.communication.phonenumbers.siprouting.aio import SipRoutingClient
+
+class DeleteSipTrunkSampleAsync(object):
+    def __init__(self):
+        connection_string = os.getenv("COMMUNICATION_SAMPLES_CONNECTION_STRING")
+        self._client = SipRoutingClient.from_connection_string(connection_string)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/samples/siprouting/delete_sip_trunk_sample_async.py:27]: The clients in samples should be demonstrated as a context manager. The async client in particular will give users warnings if it isn't used in a context manager or closed.

Comment (annatisch) [reply sdk/communication/azure-communication-phonenumbers/samples/siprouting/delete_sip_trunk_sample_async.py:27]: Ah - I see you're calling close at the bottom. It would be cleaner to have the sample demonstrate a context manager as this is generally the syntax/pattern we would recommend people use.

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample .py [hunk 11]
```diff
@@ -0,0 +1,38 @@
+# -------------------------------------------------------------------------
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample .py:1]: This file seems to have a space character in the name?

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample .py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+from azure.communication.phonenumbers.siprouting import SipRoutingClient
+
+class SetSipTrunksSample(object):
+    def __init__(self):
+        connection_string = os.getenv("COMMUNICATION_SAMPLES_CONNECTION_STRING")
+        self._client = SipRoutingClient.from_connection_string(connection_string)
+
+    def set_sip_trunks_sample(self):
+        new_trunks = os.getenv("COMMUNICATION_SAMPLES_NEW_TRUNKS")
+        sip_trunks = self._client.set_trunks(new_trunks)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample .py:30]: I didn't think `set_trunks` returned anything?

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunk_sample.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+from azure.communication.phonenumbers.siprouting import SipRoutingClient
+
+class SetSipTrunkSample(object):
+    def __init__(self):
+        connection_string = os.getenv("COMMUNICATION_SAMPLES_CONNECTION_STRING")
+        self._client = SipRoutingClient.from_connection_string(connection_string)
+
+    def set_sip_trunk_sample(self):
+        sip_trunk = os.getenv("COMMUNICATION_SAMPLES_TRUNK")
+        self._client.set_trunk(sip_trunk)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunk_sample.py:30]: Shouldn't think take an instance of SipTrunk as input?

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample_async.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+from azure.communication.phonenumbers.siprouting.aio import SipRoutingClient
+
+class SetSipTrunksSample(object):
+    def __init__(self):
+        connection_string = os.getenv("COMMUNICATION_SAMPLES_CONNECTION_STRING")
+        self._client = SipRoutingClient.from_connection_string(connection_string)
+
+    async def set_sip_trunks_sample(self):
+        new_trunks = os.getenv("COMMUNICATION_SAMPLES_NEW_TRUNKS")
+        sip_trunks = await self._client.set_trunks(new_trunks)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample_async.py:31]: This should take a list of objects as input.

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_routes_sample.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+from azure.communication.phonenumbers.siprouting import SipRoutingClient
+
+class SetSipRoutesSample(object):
+    def __init__(self):
+        connection_string = os.getenv("COMMUNICATION_SAMPLES_CONNECTION_STRING")
+        self._client = SipRoutingClient.from_connection_string(connection_string)
+
+    def set_sip_routes_sample(self):
+        new_routes = os.getenv("COMMUNICATION_SAMPLES_NEW_ROUTES")
+        sip_routes = self._client.set_routes(new_routes)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_routes_sample.py:30]: Same comment as for the `set_trunks` sample - this doesn't demonstrate how to construct the input - and shouldn't return anything.

File: sdk/communication/azure-communication-phonenumbers/README.md (hunk 1 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/README.md]: Changed to provide an example of input.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/__init__.py (hunk 2 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/__init__.py]: Removed.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/__init__.py (hunk 3 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/__init__.py]: Removed and added license header.

File: sdk/communication/azure-communication-phonenumbers/README.md (hunk 4 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/README.md:196]: Added print of result properties.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py (hunk 5 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py]: Removed as positional. The parameter is actually not used in this class, but further down the line is popped from the kwargs.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py (hunk 6 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py]: Changed. Looking at other languages, it makes the most sense to just return None, in case it's not found.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py (hunk 7 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py]: Changed to iterables.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py (hunk 8 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py]: Changed to iterables.

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/delete_sip_trunk_sample_async.py (hunk 10 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/samples/siprouting/delete_sip_trunk_sample_async.py]: Added context manager pattern.

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample .py (hunk 11 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample .py]: Fixed.

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunk_sample.py (hunk 13 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunk_sample.py]: Changed samples to either take environment variables, in case of simple inputs and provided example values, in case of structured inputs like list of routes or list of trunks.

File: sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample_async.py (hunk 14 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/samples/siprouting/set_sip_trunks_sample_async.py]: Fixed. Added example list of trunks.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+        """
+        endpoint, credential = parse_connection_str(conn_str)
+        return cls(endpoint, credential, **kwargs)
+
+    @distributed_trace
+    def get_trunk(
+        self,
+        trunk_fqdn,  # type: str
+        **kwargs  # type: Any
+    ):  # type: (...) -> SipTrunk
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py:97]: Return type should be `Optional[SipTrunk]`

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+        """Getter for single SIP trunk.
+
+        :param trunk_fqdn: FQDN of the desired SIP trunk.
+        :type trunk_fqdn: str
+        :returns: SIP trunk with specified trunk_fqdn.
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py:102]: We should elaborate that if the specified trunk is not found, None is returned.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+        :rtype: ~azure.communication.siprouting.models.SipTrunk
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py:103]:
```suggestion
        :rtype: ~azure.communication.siprouting.models.SipTrunk or None
```

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+            raise ValueError("Parameter 'trunks' must not be None.")
+
+        old_trunks = self._get_trunks_(**kwargs)
+
+        if len(old_trunks) > 0:
+            self._rest_service.patch_sip_configuration(
+                body=SipConfiguration(trunks={x.fqdn: None for x in old_trunks}), **kwargs
+            )
+        if len(trunks) > 0:
+            self._patch_trunks_(trunks, **kwargs)
```
Comment (annatisch) [inline sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py:211]: Going multiple REST calls within a single API can be problematic.
Are lines 207 and 211 mutually exclusive? Only one of these two will be called right?
Is this method idempotent if a network failure for example interrupted the patch operation and it was retried?

Is this pattern of GET -> PATCH the same pattern that's used across languages?

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py (hunk 18 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py]: Fixed.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py (hunk 17 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py]: Fixed.

File: sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py (hunk 19 above)
Comment (jiriburant) [reply sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py]: Changed the call logic, to have just one GET and one PATCH.

This pattern is used across languages, we dont have another option to wipe the trunks other than first retrieving them and then deleting. From the point of view of idempotency, this should be ok.
The operation will always retrieve the current list of trunks and overwrite them, so idempotency shouldn't be an issue.

An issue could be concurrency, but as discussed at the Arch reviews, it doesn't make sense from user perspective to update the configuration from multiple places at once, because it could always cause inconsistent configuration, so concurrency is not a concern.

**Review by annatisch — APPROVED** [review]
